querybook/server/tasks/disable_scheduled_docs.py

```diff
@@ -40,27 +87,146 @@ def get_task_owner(task, session):
+                "data_doc_title": doc.title or "Untitled",
+                "doc_url": f"{QuerybookSettings.PUBLIC_URL}/{doc.environment.name}/datadoc/{doc_id}/",
+                "disabled_reason": task_to_disable["reason"],
+            },
+            notifier_name=notifier,
+            session=session,
+        )
+
+
 @with_session
-def disable_deactivated_scheduled_docs(session=None):
-    disabled_scheduled_doc_ids = []
-    tasks = get_scheduled_datadoc_tasks(session=session)
-    for task in tasks:
-        task_owner = get_task_owner(task, session=session)
-        if task_owner.deleted:
-            update_task_schedule(task.id, commit=False, session=session, enabled=False)
-            disabled_scheduled_doc_ids.append(task.kwargs["doc_id"])
+def disable_deactivated_scheduled_docs(
+    dry_run: bool = False,
+    notifier: Optional[str] = None,
+    disable_if_inactive_owner: bool = True,
+    disable_if_failed_for_n_runs: Optional[int] = 5,
+    disable_if_no_impression_for_n_days: Optional[int] = 30,
+    session=None,
+):
+    tasks_to_disable: list[dict] = []
+    tasks_to_delete: list[int] = []
+    task_dicts = get_scheduled_datadoc_tasks(session=session)
+    for task_dict in task_dicts:
+        should_disable = None
+
+        if not check_if_doc_exists(task_dict, session=session):
+            tasks_to_delete.append(task_dict["id"])
+            continue
+
+        if disable_if_inactive_owner and check_task_inactive_owner(task_dict, session):
+            should_disable = DisablePolicy.INACTIVE_OWNER
+
+        if (
+            should_disable is None
+            and disable_if_failed_for_n_runs is not None
+            and check_task_failed_for_n_runs(
+                task_dict, disable_if_failed_for_n_runs, session
+            )
+        ):
+            should_disable = DisablePolicy.FAILED_FOR_N_RUNS
+
+        if (
+            not should_disable
+            and disable_if_no_impression_for_n_days is not None
+            and check_task_no_impression_for_n_days(
+                task_dict, disable_if_no_impression_for_n_days, session
+            )
+        ):
+            should_disable = DisablePolicy.NO_IMPRESSION_FOR_N_DAYS
+
+        if should_disable is not None:
+            tasks_to_disable.append(
+                {
+                    "doc_id": task_dict["kwargs"]["doc_id"],
+                    "task_id": task_dict["id"],
+                    "reason": disable_policy_to_reason(
+                        should_disable,
+                        disable_if_failed_for_n_runs,
+                        disable_if_no_impression_for_n_days,
+                    ),
+                    "task_owner_uid": task_dict["kwargs"].get("user_id"),
+                }
+            )
+
+    if not dry_run:
+        for task_id in tasks_to_delete:
+            delete_task_schedule(task_id, commit=False, session=session)
+
+        for task_to_disable in tasks_to_disable:
+            update_task_schedule(
```

Comment:
do we need to save the disable reason back to db for  the schedule?

Reply:
I don't have a good place to put it, but I was thinking to leverage the event logs you built. What do you think?

Reply:
how about setting the task schedule as disabled instead of deleting it? and put the disable reason in the kwargs?

Reply:
otherwise, with deleting, after they fixed the data doc, they'll have to recreate the schedule and fill in all the notification, exporters again.

Reply:
These task schedules are getting disabled not deleted. The deletion of task schedule only happens to docs that are deleted, in this case no action can be done on them.